protocols/smtp/src/main/java/org/apache/james/protocols/smtp/core/AbstractSenderAuthIdentifyVerificationRcptHook.java

```diff
@@ -45,28 +47,36 @@
     @Override
     public HookResult doRcpt(SMTPSession session, MaybeSender sender, MailAddress rcpt) {
         if (session.getUser() != null) {
-            String authUser = (session.getUser()).toLowerCase(Locale.US);
-            MailAddress senderAddress = (MailAddress) session.getAttachment(
-                    SMTPSession.SENDER, ProtocolSession.State.Transaction);
-            String username = retrieveSender(sender, senderAddress);
+            MaybeSender senderAddress = (MaybeSender) session.getAttachment(SMTPSession.SENDER, ProtocolSession.State.Transaction);
             
             // Check if the sender address is the same as the user which was used to authenticate.
             // Its important to ignore case here to fix JAMES-837. This is save todo because if the handler is called
             // the user was already authenticated
-            if ((senderAddress == null)
-                || (!authUser.equalsIgnoreCase(username))
-                || (!isLocalDomain(senderAddress.getDomain()))) {
+            if (isAnonymous(sender)
+                || !hasRightUsername(sender, session)
+                || !belongsToLocalDomain(senderAddress)) {
                 return INVALID_AUTH;
             }
         }
         return HookResult.DECLINED;
     }
 
-    public String retrieveSender(MaybeSender sender, MailAddress senderAddress) {
-        if (senderAddress != null && !sender.isNullSender()) {
-            return getUser(senderAddress);
-        }
-        return null;
+    private boolean isAnonymous(MaybeSender maybeSender) {
+        return maybeSender == null || maybeSender.isNullSender();
+    }
+
+    private boolean hasRightUsername(MaybeSender maybeSender, SMTPSession session) {
+        Preconditions.checkArgument(!maybeSender.isNullSender());
+
+        String authUser = session.getUser().toLowerCase(Locale.US);
+        String username = getUser(maybeSender.get());
```

Comment:
do we need to convert string `toLowerCase()` here?

Reply:
Well, most of this logic is not covered by tests, changing it too much do not seems like a good idea.

Reply:
https://ci.linagora.com/linagora/lgs/openpaas/james/issues/961